core/src/main/java/org/infinispan/stream/impl/LockStreamImpl.java

```diff
@@ -0,0 +1,231 @@
+import org.infinispan.util.concurrent.locks.KeyAwareLockPromise;
+import org.infinispan.util.concurrent.locks.LockManager;
+import org.infinispan.util.function.SerializablePredicate;
+
+/**
+ * Lock Stream implementation that locks a key using the {@link LockManager} before and after executing the various
+ * code.
+ * <p>
+ * This implementation doesn't work properly when using an optimistic transactional cache. Care should be made to prevent
+ * that usage if possible.
+ * @author wburns
+ * @since 9.0
+ */
+public class LockStreamImpl<K, V> implements LockedStream<K, V> {
+   private final CacheStream<CacheEntry<K, V>> realStream;
+   private final Predicate<? super CacheEntry<K, V>> predicate;
+
+   public LockStreamImpl(CacheStream<CacheEntry<K, V>> realStream) {
+      Objects.nonNull(realStream);
+      this.realStream = realStream;
+      this.predicate = null;
+   }
+
+   private LockStreamImpl(CacheStream<CacheEntry<K, V>> realStream, Predicate<? super CacheEntry<K, V>> predicate) {
+      this.realStream = realStream;
+      this.predicate = predicate;
+   }
+
+   private LockedStream<K, V> newOrReuse(CacheStream<CacheEntry<K, V>> resultingStream) {
+      if (resultingStream == realStream) {
+         return this;
+      } else {
+         return new LockStreamImpl<>(resultingStream, predicate);
+      }
+   }
+
+   @Override
+   public LockedStream<K, V> filter(final Predicate<? super CacheEntry<K, V>> predicate) {
+      Objects.nonNull(predicate);
+      Predicate<? super CacheEntry<K, V>> usedPredicate;
+      if (this.predicate != null) {
+         usedPredicate = (SerializablePredicate<? super CacheEntry<K, V>>) e -> this.predicate.test(e) && predicate.test(e);
+      } else {
+         usedPredicate = predicate;
+      }
+      return new LockStreamImpl<>(realStream, usedPredicate);
+   }
+
+   @Override
+   public void forEach(BiConsumer<Cache<K, V>, ? super CacheEntry<K, V>> biConsumer) {
+      realStream.forEach(new CacheEntryConsumer<>(biConsumer, predicate));
+   }
+
+   @Override
+   public BaseCacheStream sequentialDistribution() {
+      return newOrReuse(realStream.sequentialDistribution());
+   }
+
+   @Override
+   public BaseCacheStream parallelDistribution() {
+      return newOrReuse(realStream.parallelDistribution());
+   }
+
+   @Override
+   public BaseCacheStream filterKeySegments(Set<Integer> segments) {
+      return newOrReuse(realStream.filterKeySegments(segments));
+   }
+
+   @Override
+   public BaseCacheStream filterKeys(Set<?> keys) {
+      return newOrReuse(realStream.filterKeys(keys));
+   }
+
+   @Override
+   public BaseCacheStream distributedBatchSize(int batchSize) {
+      return newOrReuse(realStream.distributedBatchSize(batchSize));
+   }
+
+   @Override
+   public BaseCacheStream segmentCompletionListener(SegmentCompletionListener listener) {
+      throw new UnsupportedOperationException("LockStream doesn't support completion listener");
```

Comment:
Please add comment: is it just not-implemented-yet, or is something principially blocking impl?